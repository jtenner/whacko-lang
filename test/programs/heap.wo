export fn _start(): void {
  let a = heap.malloc(usize(100));
  heap.free(a);
}