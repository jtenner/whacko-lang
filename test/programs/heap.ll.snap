; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

declare void @__whacko_gc_initialize()

declare ptr @__whacko_gc_alloc(i32, i32)

declare void @__whacko_gc_store_field_ptr(i32, ptr, ptr, ptr)

define void @_start() {
entry:
  call void @__whacko_gc_initialize()
  %"alloca~0" = alloca i32, align 4
  %"471~inst" = tail call ptr @malloc(i32 mul (i32 ptrtoint (ptr getelementptr (i8, ptr null, i32 1) to i32), i32 100))
  %"472~inst" = ptrtoint ptr %"471~inst" to i32
  store i32 %"472~inst", ptr %"alloca~0", align 4
  %"474~inst" = load i32, ptr %"alloca~0", align 4
  %"476~inst" = inttoptr i32 %"474~inst" to ptr
  tail call void @free(ptr %"476~inst")
  ret void
}

declare noalias ptr @malloc(i32) #0

declare void @free(ptr) #1

declare void @__whacko_gc_visit(ptr)

define void @__visit_whacko_object(ptr %0) {
"entry~480":
  unreachable
}

attributes #0 = { "alloc-family"="malloc" "allockind"="alloc,zeroed" "allocsize"="0" }
attributes #1 = { "alloc-family"="malloc" "allockind"="free" }
